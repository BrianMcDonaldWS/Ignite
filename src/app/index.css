.App {
  display: flex;
}

footer {
  background-color: #b9b6b6;
}

code {
  display: block;
  padding: 0.5em;
}

.noMargin {
  margin-left: 0;
  margin-right: 0;
}
