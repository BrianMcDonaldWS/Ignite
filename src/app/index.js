import React from 'react';
import { hydrate, render } from 'react-dom';
import { BrowserRouter, Route } from 'react-router-dom';

import withGlobalConfig from './components/WithGlobalConfig';
import MarkdownProvider from './components/MarkdownProvider';

const MarkdownProviderWithConfig = withGlobalConfig(MarkdownProvider);
const rootElement = document.getElementById('index');

const IgniteApp = (
  <BrowserRouter>
    <Route component={props => <MarkdownProviderWithConfig {...props} />} />
  </BrowserRouter>
);

if (rootElement.hasChildNodes()) {
  hydrate(IgniteApp, rootElement);
} else {
  render(IgniteApp, rootElement);
}
