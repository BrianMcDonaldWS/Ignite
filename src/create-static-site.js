#!/usr/bin/env node

import fs from 'fs';
import url from 'url';
import path from 'path';
import http from 'http';
import mkdirp from 'mkdirp';
import root from 'root-path';
import cpy from 'cpy';

import handler from 'serve-handler';
import puppeteer from 'puppeteer';

export const startServer = options =>
  http.createServer((request, response) =>
    handler(request, response, {
      public: options.dst,
      cleanUrls: false,
      rewrites: [
        {
          source: path.join(options.baseURL, '**'),
          destination: path.join(options.baseURL, 'index.html')
        }
      ]
    })
  );

const getLinks = async page => {
  const anchors = await page.evaluate(() =>
    [...document.querySelectorAll('a')].map(anchor => anchor.href)
  );

  const iframes = await page.evaluate(() =>
    [...document.querySelectorAll('iframe')].map(iframe => iframe.src)
  );

  return anchors.concat(iframes);
};

const writeHtml = async (page, filePath, options) => {
  const content = await page.content();

  filePath = path.join(root(), options.dst, 'static', filePath);

  if (filePath.endsWith('.html')) {
    mkdirp.sync(path.dirname(filePath));
    fs.writeFileSync(filePath, content);
  }
};

export default function createStaticSite(options) {
  const server = startServer(options);
  const log = (...string) => options.log && console.warn(...string);

  return new Promise(resolve =>
    server.listen(options.port, async () => {
      log(`server is listening on ${options.port}`);
      log('Generating static page for:');

      const rootIndex = path
        .join(options.baseURL, options.index)
        .replace('.md', '');
      const browser = await puppeteer.launch();
      const linksVisited = new Set();
      const linksToVisit = new Set([
        url.resolve(`http://localhost:${options.port}`, options.baseURL)
      ]);

      await cpy(
        ['*.css', '*.js', '**/*.{png,svg,gif,jpg,jpeg}'],
        path.join(root(), options.dst, 'static', options.baseURL),
        {
          parents: true,
          cwd: path.join(root(), options.dst, options.baseURL)
        }
      );

      async function processLink() {
        const [link] = [...linksToVisit];

        if (!link) {
          const page = await browser.newPage();
          await page.setViewport({ width: 1200, height: 1000 });
          await page.goto(
            url.resolve(`http://localhost:${options.port}`, rootIndex)
          );
          await writeHtml(page, rootIndex, options);
          browser.close();
          server.close();
          resolve();
          return;
        }

        log('**', link);
        linksToVisit.delete(link);
        linksVisited.add(link);

        const page = await browser.newPage();
        await page.setViewport({ width: 1200, height: 1000 });
        await page.goto(link);

        const newLinks = await getLinks(page);
        newLinks
          .filter(
            link =>
              link !== '' &&
              link.includes(`localhost:${options.port}`) &&
              link.includes('.html')
          )
          .forEach(link => {
            const [newLink] = link.split('#');

            if (!linksVisited.has(newLink)) {
              linksToVisit.add(newLink);
            }
          });

        const file = url.parse(link).pathname;
        if (file !== rootIndex) {
          await writeHtml(page, file, options);
        }
        page.close();
        processLink();
      }

      return processLink();
    })
  );
}
