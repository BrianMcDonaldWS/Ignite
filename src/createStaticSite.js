#!/usr/bin/env node

import fs from 'fs';
import url from 'url';
import path from 'path';
import http from 'http';
import mkdirp from 'mkdirp';
import root from 'root-path';

import handler from 'serve-handler';
import puppeteer from 'puppeteer';

const startServer = options =>
  http.createServer((request, response) => {
    return handler(request, response, {
      cleanUrls: false,
      rewrites: [
        {
          source: path.join(options.baseURL, '**'),
          destination: path.join(options.baseURL, 'index.html')
        }
      ]
    });
  });

const getLinks = async page => {
  const anchors = await page.evaluate(() =>
    [...document.querySelectorAll('a')].map(anchor => anchor.href)
  );

  const iframes = await page.evaluate(() =>
    [...document.querySelectorAll('iframe')].map(iframe => iframe.src)
  );

  return anchors.concat(iframes);
};

const writeHtml = async (page, filePath) => {
  const content = await page.content();

  filePath = path.join(root(), filePath);

  if (filePath.endsWith('.html')) {
    mkdirp.sync(path.dirname(filePath));
    fs.writeFileSync(filePath, content);
  }
};

export default function createStaticSite(options) {
  const server = startServer(options);

  return new Promise(resolve =>
    server.listen(options.port, async () => {
      console.warn(`server is listening on ${options.port}`);
      console.warn('Generating static page for:');

      const rootIndex = path
        .join(options.baseURL, options.index)
        .replace('.md', '.html');
      const browser = await puppeteer.launch();
      const linksVisited = new Set();
      const linksToVisit = new Set([
        url.resolve(`http://localhost:${options.port}`, options.baseURL)
      ]);

      async function processLink() {
        const [link] = [...linksToVisit];

        if (!link) {
          const page = await browser.newPage();
          await page.setViewport({ width: 1200, height: 1000 });
          await page.goto(
            url.resolve(`http://localhost:${options.port}`, rootIndex)
          );
          await writeHtml(page, rootIndex);
          browser.close();
          server.close();
          resolve();
          return;
        }

        console.warn('**', link);
        linksToVisit.delete(link);
        linksVisited.add(link);

        const page = await browser.newPage();
        await page.setViewport({ width: 1200, height: 1000 });
        await page.goto(link);

        const newLinks = await getLinks(page);
        newLinks
          .filter(
            link =>
              link !== '' &&
              link.includes(`localhost:${options.port}`) &&
              link.includes('.html')
          )
          .forEach(link => {
            const [newLink] = link.split('#');

            if (!linksVisited.has(newLink)) {
              linksToVisit.add(newLink);
            }
          });

        const file = url.parse(link).pathname;
        if (file !== rootIndex) {
          await writeHtml(page, file);
        }
        page.close();
        processLink();
      }

      return processLink();
    })
  );
}
