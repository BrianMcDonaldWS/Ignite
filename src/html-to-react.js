import path from 'path';
import { getOptions } from 'loader-utils';

export default function(source) {
  const options = getOptions(this);
  const pathToMarkdown = path.relative(options.src, this.resourcePath);
  const isIndex = this.resourcePath.includes(path.join(options.src, options.index));

  return `
    import ignite from 'ignite';

    const markDownPage = () => (
      <div>
        ${source}
      </div>
    );
    
    export default ignite('${pathToMarkdown}', markDownPage, ${isIndex});
  `;
}
