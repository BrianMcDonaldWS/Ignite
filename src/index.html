<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">

  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.12/css/all.css" integrity="sha384-G0fIWCsCzJIMAVNQPfjH08cyYaUtMwjJwqiRKxxE/rx96Uroj1BtIQ6MLJuheaO9" crossorigin="anonymous">
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/<%= htmlWebpackPlugin.options.codeStyle %>.min.css">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.1/css/bulma.min.css">
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/bulmaswatch/0.6.2/<%= htmlWebpackPlugin.options.bulmaTheme %>/bulmaswatch.min.css">

  <title>Ignite</title>
</head>

<body>
  <div id="index" />
</body>

</html>