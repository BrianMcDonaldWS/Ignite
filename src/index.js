import React from 'react';
import ReactDOM from 'react-dom';

const markdown = {};

const Index = props => {
  console.log(props);
  return <div>Hello React!</div>;
};

ReactDOM.render(<Index />, document.getElementById('index'));

export default function registerMarkdown(path, markdownInJS) {
  markdown[path] = markdownInJS;
  console.log('registerMarkdown', path);
  return markdownInJS;
}
