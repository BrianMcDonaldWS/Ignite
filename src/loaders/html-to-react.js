import path from 'path';
import { getOptions } from 'loader-utils';

function getLink(source, index = 0) {
  const linkIndex = source.indexOf('<a h', index);

  if (linkIndex === -1) {
    return '';
  }

  const hrefIndex = source.indexOf('href="', linkIndex);
  const startIndex = source.indexOf('/', hrefIndex) + 1;
  const endIndex = source.indexOf('"', startIndex);

  return source.substring(startIndex, endIndex);
}

function addActive(source, link, firstLink) {
  source = source.replace(
    new RegExp('<a h'),
    `<a
      className={
        '/${link}' === props.currentPage ||
        ('${firstLink}' === '${link}' && '/' === props.currentPage)
          ? 'is-active'
          : null
      }
      h`
  );

  return source;
}

function addActiveAll(source, firstLink) {
  let nextLink = getLink(source);

  while (nextLink !== '') {
    source = addActive(source, nextLink, firstLink);
    nextLink = getLink(source);
  }

  return source;
}

function index(source, pathToMarkdown) {
  const firstLink = getLink(source);

  // Some of the curlies need to stay to pass props to the plugin component
  source = source.replace(new RegExp('!{', 'g'), '__CURLY_LEFT__');
  source = source.replace(new RegExp('!}', 'g'), '__CURLY_RIGHT__');
  source = addActiveAll(source, firstLink);
  source = source.replace(
    new RegExp('<ul>', 'g'),
    '<ul className="menu-list">'
  );
  source = source.replace(new RegExp('<p>', 'g'), '<p className="menu-label">');
  source = source.replace(new RegExp('__CURLY_LEFT__', 'g'), '{');
  source = source.replace(new RegExp('__CURLY_RIGHT__', 'g'), '}');

  return `
    import { registerMarkdown } from 'ignite';
    import makeClass from 'classnames';

    function markDownPage(props) {
      return (
        <aside className={makeClass('menu', props.className)}>
          ${source}
        </aside>
      );
    }
    
    export default registerMarkdown('${pathToMarkdown}', markDownPage, true, '${firstLink}');
  `;
}

function replaceAt(input, search, replace, start) {
  return (
    input.slice(0, start) +
    input.slice(start, start + search.length).replace(search, replace) +
    input.slice(start + search.length)
  );
}

function insertBreaks(source) {
  let preTeg = source.indexOf('<pre>');

  while (preTeg !== -1) {
    let endPreTag = source.indexOf('</pre>', preTeg);
    let newLine = source.indexOf('\n', preTeg);

    do {
      newLine = source.indexOf('\n', preTeg);
      endPreTag = source.indexOf('</pre>', preTeg);
      source = replaceAt(source, '\n', '<br />', newLine);
    } while (
      newLine < endPreTag &&
      newLine !== source.indexOf('\n', preTeg) &&
      newLine !== -1
    );

    preTeg = source.indexOf('<pre>', endPreTag);
  }

  return source;
}

const regexIndexOf = function(string, regex, startpos) {
  const indexOf = string.substring(startpos || 0).search(regex);
  return indexOf >= 0 ? indexOf + (startpos || 0) : indexOf;
};

const replaceIdLinks = source => {
  let isTag = /<a href="#(?!\/)[\S]+/;
  let linkOnPage = regexIndexOf(source, isTag);

  while (linkOnPage !== -1) {
    source = replaceAt(source, '<a href="#', '<Link to="#', linkOnPage);
    source = replaceAt(
      source,
      '</a>',
      '</Link>',
      source.indexOf('</a>', linkOnPage)
    );
    linkOnPage = regexIndexOf(source, isTag, linkOnPage);
  }

  isTag = /<a className="fas fa-hashtag headerLink" href="#(?!\/)[\S]+/;
  linkOnPage = regexIndexOf(source, isTag);
  console.log(linkOnPage);
  while (linkOnPage !== -1) {
    source = replaceAt(
      source,
      '<a className="fas fa-hashtag headerLink" href="#',
      '<Link className="fas fa-hashtag headerLink" to="#',
      linkOnPage
    );
    source = replaceAt(
      source,
      '</a>',
      '</Link>',
      source.indexOf('</a>', linkOnPage)
    );
    linkOnPage = regexIndexOf(source, isTag, linkOnPage);
  }

  return source;
};

export function sanitizeJSX(source) {
  if (source.includes('<pre>')) {
    source = insertBreaks(source);
  }

  // Some of the curlies need to stay to pass props to the plugin component
  source = source.replace(new RegExp('!{', 'g'), '__CURLY_LEFT__');
  source = source.replace(new RegExp('!}', 'g'), '__CURLY_RIGHT__');

  // Don't break the JSX
  // source = source.replace(new RegExp('`', 'g'), '\\`');
  source = source.replace(new RegExp('{', 'g'), '&#123;');
  source = source.replace(new RegExp('}', 'g'), '&#125;');

  // Need to escape br for jsx
  source = source.replace(new RegExp('<br>', 'g'), '<br/>');

  source = source.replace(new RegExp('__CURLY_LEFT__', 'g'), '{');
  source = source.replace(new RegExp('__CURLY_RIGHT__', 'g'), '}');

  // Uppercase to use as react component
  source = source.replace(new RegExp('pluginprovider', 'g'), 'PluginProvider');

  // React uses className
  source = source.replace(new RegExp('class=', 'g'), 'className=');
  source = replaceIdLinks(source, /<a href="#(?!\/)[\S]+/);

  return source;
}

export default function(source) {
  const options = getOptions(this);
  const pathToMarkdown = path.relative(options.src, this.resourcePath);
  const isIndex = this.resourcePath.includes(
    path.join(options.src, options.index)
  );

  if (isIndex) {
    return index(source, pathToMarkdown);
  }

  source = sanitizeJSX(
    source.replace(
      new RegExp('highlighted-line', 'g'),
      'highlighted-line hero is-primary'
    )
  );

  return `
    import React from 'react';
    import { registerMarkdown } from 'ignite';
    import { Link } from 'react-router-dom';

    const PluginProvider = ({plugins, name, options, children}) => {
      let Plugin = plugins[name];
    
      if (!Plugin) {
        return <div />;
      }

      Plugin = Plugin.component;

      return <Plugin options={options.options} children={children} {...options.properties} />;
    };

    const markDownPage = props => (
      <div className={props.className}>
        <section>
          ${source}
        </section>
      </div>
    );
    
    export default registerMarkdown('${pathToMarkdown}', markDownPage);
  `;
}
