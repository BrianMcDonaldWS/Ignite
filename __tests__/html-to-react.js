import { insertInString, sanitizeJSX } from '../src/html-to-react';

test('insertString', () => {
  expect(insertInString('1245', '3', 2)).toBe('12345');
});

test('sanitizeJSX', () => {
  expect(
    sanitizeJSX(`
    <code>
      function foo() {
        console.log('foo');
      }
    </code>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeJSX(`
    <code>
      function foo() {
        console.log('foo');
      }
    </code>
    <code>
      function bar() {
        console.log('bar');
      }
    </code>
`)
  ).toMatchSnapshot();

  expect(
    sanitizeJSX(`
    {
      some: 'object
    }
  `)
  ).toBe();
});
