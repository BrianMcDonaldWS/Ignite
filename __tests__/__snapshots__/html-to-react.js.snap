// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanitizeJSX 1`] = `
"
    <code>
      function foo() {
        console.log('foo');
      }
    </code>
  "
`;

exports[`sanitizeJSX 2`] = `
"
    <code>
      function foo() {
        console.log('foo');
      }
    </code>
    <code>
      function bar() {
        console.log('bar');
      }
    </code>
"
`;

exports[`sanitizeJSX 3`] = `
"
    \\"{
      some: 'object
    \\"}
  "
`;
